providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,782 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "prov/hybrid_pkey.h"
+#include "hybrid_kmgmt_local.h"
+
+#include <string.h>
+#include <openssl/core_names.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/proverr.h>
+#include "crypto/evp.h"
+#include "prov/providercommon.h"
+#include "prov/provider_ctx.h"
+#include "internal/common.h"
+#include "internal/e_os.h"
+#include "internal/params.h"
+#include "internal/param_names.h"
+#include "internal/refcount.h"
+
+HYBRID_PKEY *ossl_hybrid_kmgmt_new(void *provctx,
+                                   const HYBRID_ALG_INFO *info)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)OPENSSL_zalloc(sizeof(*key));
+    unsigned int i;
+
+    if (key == NULL)
+        return NULL;
+    key->libctx = PROV_LIBCTX_OF(provctx);
+    key->info = info;
+    if (!CRYPTO_NEW_REF(&key->references, 1))
+        goto err;
+    INIT_ACCUMULATE_HYBRID_NUMBERS(key);
+    for (i = 0; i < key->info->num_algs; i++) {
+        if ((key->keys[i] = EVP_PKEY_new()) == NULL)
+            goto err;
+        ACCUMULATE_HYBRID_NUMBERS(key, i);
+    }
+    return key;
+
+ err:
+    ossl_hybrid_kmgmt_free(key);
+    return NULL;
+}
+
+void ossl_hybrid_pkey_free(HYBRID_PKEY *key)
+{
+    unsigned int i;
+    int r;
+
+    if (key == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&key->references, &r);
+    REF_PRINT_COUNT("HYBRID", key);
+    if (r > 0)
+        return;
+    REF_ASSERT_ISNT(r < 0);
+    CRYPTO_FREE_REF(&key->references);
+
+    for (i = 0; i < key->info->num_algs; i++)
+        EVP_PKEY_free(key->keys[i]);
+    OPENSSL_free(key->propq);
+    OPENSSL_free(key);
+}
+
+int ossl_hybrid_pkey_up_ref(HYBRID_PKEY *key)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&key->references, &i) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("HYBRID", key);
+    REF_ASSERT_ISNT(i < 2);
+    return i > 1 ? 1 : 0;
+}
+
+void ossl_hybrid_kmgmt_free(void *vkey)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+
+    ossl_hybrid_pkey_free(key);
+}
+
+void *ossl_hybrid_kmgmt_dup(const void *vold, ossl_unused int selection)
+{
+    const HYBRID_PKEY *old = (HYBRID_PKEY *)vold;
+    HYBRID_PKEY *key;
+    unsigned int i;
+
+    if (old == NULL
+            || (key = (HYBRID_PKEY *)OPENSSL_zalloc(sizeof(*key))) == NULL)
+        return NULL;
+    if (old->propq != NULL && (key->propq = strdup(old->propq)) == NULL)
+        goto err;
+    for (i = 0; i < key->info->num_algs; i++)
+        if ((key->keys[i] = EVP_PKEY_dup(old->keys[i])) == NULL)
+            goto err;
+    key->info = old->info;
+    key->key_length = old->key_length;
+    key->bits = old->bits;
+    key->security_bits = old->security_bits;
+    return key;
```

Comment:
What about the refcount? The duplicated key should have a refcount of 1.

Reply:
Good catch.